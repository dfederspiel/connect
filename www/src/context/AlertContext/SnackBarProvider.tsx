import React, { useState, useEffect, createContext, useContext } from 'react';
import Snackbar from '@material-ui/core/Snackbar';
import { SnackBarType } from './SnackBarType';

const snackBarContext = createContext<SnackBarType>(null as unknown as SnackBarType);
interface SnackBarProviderProps {
  children: JSX.Element;
  timeout: number;
}

// export const snackBarContext = createContext<SnackBarType>({message: '', updateMessage: () => {}});
export const SnackBarProvider = ({
  children,
  timeout,
}: SnackBarProviderProps): JSX.Element => {
  const [message, setMessage] = useState('');
  const updateMessage = (m: string) => {
    setMessage(m);
    setOpen(true);
  };
  const [open, setOpen] = useState(true);

  useEffect(() => {
    if (message.length > 0 && open) {
      const timer = setTimeout(() => setOpen(false), timeout);
      return () => clearTimeout(timer);
    }
  });

  return (
    <snackBarContext.Provider value={{ open, message, updateMessage }}>
      {children}
      {message.length > 0 && (
        <Snackbar
          open={open}
          anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
          message={message}
        />
      )}
    </snackBarContext.Provider>
  );
};
// Hook for child components to get the snack object ...
// ... and re-render when it changes.
export const useSnacks = (): SnackBarType => useContext(snackBarContext);
