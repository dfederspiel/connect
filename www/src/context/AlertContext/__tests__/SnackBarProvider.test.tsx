import { waitFor } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import { SnackBarProvider, SnackBarProviderProps, useSnacks } from '../SnackBarProvider';

describe('the toast provider', () => {
  const wrapper = ({ children }: SnackBarProviderProps) => (
    <SnackBarProvider timeout={100}>{children}</SnackBarProvider>
  );
  it('exists', async () => {
    const { result } = renderHook(() => useSnacks(), { wrapper });
    act(() => {
      result.current.updateMessage('test');
    });
    expect(result.current.message).toEqual('test');
    expect(result).toMatchSnapshot();
    await waitFor(() => {
      expect(result.current.open).toBeFalsy();
    });
  });
});
