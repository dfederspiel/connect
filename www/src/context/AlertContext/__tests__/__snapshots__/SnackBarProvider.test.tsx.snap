// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the toast provider exists 1`] = `
Object {
  "all": Array [
    Object {
      "message": "",
      "open": true,
      "updateMessage": [Function],
    },
    Object {
      "message": "test",
      "open": true,
      "updateMessage": [Function],
    },
  ],
  "current": Object {
    "message": "test",
    "open": true,
    "updateMessage": [Function],
  },
  "error": undefined,
}
`;
